### ISSUE TITLE
3.0.1 incorrectly deleting required imports from typescript files

### ISSUE THREAD

**Issue by tbartley** [simonhaenisch/prettier-plugin-organize-imports#64]
Hi,

The 3.0.1 update seems to be incorrectly deleting required imports.

The following code demonstrates the issue:

If I have TypeScript:

```typescript
import { NormalizedCacheObject } from 'apollo-cache-inmemory'
import AWSAppSyncClient from 'aws-appsync'

export class Foo {
  public constructor(private readonly appSyncClient: AWSAppSyncClient<NormalizedCacheObject>) {}
  
  getClient(): any {
    return this.appSyncClient
  }
}
```

with `package.json`

```json
{
  "name": "scratch",
  "version": "1.0.0",
  "main": "index.js",
  "license": "UNLICENSED",
  "private": true,
  "scripts": {
    "prettier": "prettier --list-different \"**/*.{json,js,ts,yml,}\""
  },
  "resolutions": {
    "apollo-cache-inmemory": "^1.6.6"
  },
  "devDependencies": {
    "prettier": "^2.7.1",
    "prettier-plugin-organize-imports": "^3.0.1",
    "typescript": "^4.7.4"
  },
  "dependencies": {
    "apollo-cache-inmemory": "^1.6.6",
    "aws-appsync": "^4.1.7"
  }
}
```

then the import:

```typescript 
import AWSAppSyncClient from 'aws-appsync'
```

is flagged as needing to be deleted.

If I use 3.0.0 then the import is not flagged.

**Comment by simonhaenisch** [comment]
Hey thanks for raising this, however I'm failing to reproduce it 🤔 

I've added the following test case which passes.

```js
test('does not remove default imports used as a type', (t) => {
	const code = `import { NormalizedCacheObject } from "apollo-cache-inmemory";
import AWSAppSyncClient from "aws-appsync";

export class Foo {
  public constructor(
    private readonly appSyncClient: AWSAppSyncClient<NormalizedCacheObject>
  ) {}

  getClient(): any {
    return this.appSyncClient;
  }
}
`;

	const formattedCode = prettify(code);

	t.is(formattedCode, code);
});
```

Can you share your `tsconfig.json` as well maybe?

**Comment by tbartley** [comment]
Of course - I forgot to do that, sorry:

```json
{
  "compilerOptions": {
    "module": "commonjs",
    "declaration": false,
    "strict": true,
    "removeComments": true,
    "lib": ["ES2020"],
    "allowSyntheticDefaultImports": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "skipLibCheck": true,
    "target": "ES2020",
    "sourceMap": true,
    "allowJs": false,
    "importHelpers": true,
    "baseUrl": ".",
    "outDir": "./build",
    "downlevelIteration": true,
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "typeRoots": ["./node_modules/@types"]
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules"]
}
```

**Comment by Shinigami92** [comment]
I encounter a similar bug (so similar I don't want to open a separate issue for now)

Using `PascalCase` components in `vue setup` and the **components** are only used in the `template pug` section but now get removed from the imports.
Try to tackle this bug first and lets see if this fixes my bug also, if not I will open its own issue with a reproduction.

**Comment by simonhaenisch** [comment]
@Shinigami92 you're not talking about what was happening in #59 (removes imports if components are kebab-cased) though, right?

**Comment by Shinigami92** [comment]
> @Shinigami92 you're not talking about what was happening in #59 (removes imports if components are kebab-cased) though, right?

I'm not sure 🤔 I use `PascalCase`

**Comment by askielboe** [comment]
We have the same issue trying to upgrade from `3.0.0` to `3.0.1`. Here is a minimal reproduction:

1. Create an app using `npx create-react-app organize-bug --template typescript`

2. `src/models/Tip.ts`

```
export type Tip = {
  text: string
}
```

3. `src/App.tsx`

```
import { Tip as TipModel } from "./models/Tip"

export const Tip = () => {
    const logTip = (tip: TipModel) => console.log(tip)
    logTip({text: "hello world"})
}
```

4. Running prettier with `prettier-plugin-organize-imports` removes `import { Tip as TipModel } from "./models/Tip"`:

```
npx prettier --check .
Checking formatting...
[warn] src/App.tsx
[warn] Code style issues found in the above file. Forgot to run Prettier?
```

**Comment by simonhaenisch** [comment]
Thanks for that repro @askielboe, I can reproduce it now.

**Comment by simonhaenisch** [comment]
Hey just released `3.0.2` which should fix this, can you test it out please?

**Comment by Shinigami92** [comment]
Will do on monday

**Comment by simonhaenisch** [comment]
Btw, to not get the `React` import removed as unused, you'll have to change the `jsx` option in tsconfig to `react` (see [TS handbook](https://www.typescriptlang.org/docs/handbook/jsx.html#basic-usage)). Or, if you leave it as `react-jsx`, it's actually be fine to have the import removed because it's really not needed. I've tried and it builds and works fine without the `React` import.

**Comment by tbartley** [comment]
Hey @simonhaenisch - thanks for taking the time to have a look at this. This seems to have fixed the issue for me for my repro. However when I add 3.0.2 to my real project running prettier now takes a very long time (10s of minutes instead of seconds).


I wonder if the `ts.ScriptSnapshot.fromString(content)` results can be cached by filePath? I tried naively doing this using a local build of `prettier-plugin-organize-imports` but didn't seem to make much difference.

**Comment by askielboe** [comment]
Thanks for the update! The bug is gone I however see the same issue at @tbartley in that prettier is now extremely slow. We have about 50,000 lines of TypeScript to check and prettier timed out in CI after 10 minutes. Usually it takes around 10 seconds.

**Comment by Shinigami92** [comment]
Can confirm, it's around 10x - 20x times slower per file

**Comment by simonhaenisch** [comment]
Thanks for replying, the performance regression is being tracked in #67 now. Should be able to get a fix out today.

**Comment by arminrosu** [comment]
@simonhaenisch I had a similar issue in 3.0.1:

```ts
import type LoggerInterface from './LoggerInterface'

export default function CreateProxy(forwardUrl: string, logger: LoggerInterface) {
  const proxy = rocky()
```

The `import type` line was being removed. It's a variation on askielbo's issue. Thought I'd mention it, maybe you want to add it to your tests.

3.0.2 fixed the issue. Thanks!

**Comment by simonhaenisch** [comment]
The performance regression should be fixed in `3.0.3` if you want to try it out!

@arminrosu thanks for the repro however I couldn't reproduce it in an isolated file, i think it was related to the fact that I added some methods to the language service host that made it more project-aware.

**Comment by arminrosu** [comment]
@simonhaenisch ah, ok. Do you want me to work on a reproducible example? I guess it might have to do with my tsconfig.json.

**Comment by tbartley** [comment]
My issue is completely resolved with good performance with 3.0.3 - thamks @simonhaenisch !
